### PR TITLE
Add AI/ML engineer challenge

### PR DESCRIPTION
**Pull request by r0mant**

### CODE REVIEW

**Review by benarent — COMMENTED** [review]
This challenge is good. It's maybe a bit Go specific, but the challenge isn't too dogmatic on implementation.

File: challenges/ai-ml/challenge-1.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+  the Go language.
+* Tests exist for happy path and error scenarios for key components of the challenge.
+* Make sure builds are reproducible.
+* Ensure error handling and error reporting is consistent. The system should
+  report clear errors and not crash under non-critical conditions.
+* Avoid concurrency and networking errors. Most of the issues we've seen in
+  production are related to data races, networking error handling or goroutine
+  leaks. We will be looking for those errors in your code.
+* Security. Use strong authentication and a simple, but secure, authorization scheme.
+  Set up the strongest transport encryption you can. Test it.
```
Comment (tigrato) [inline challenges/ai-ml/challenge-1.md:158]:
```suggestion
```

File: challenges/ai-ml/challenge-1.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+a brief doc that covers the edge cases and design approach.
+
+Please submit the design document and all code in a GitHub repository. Public
+or private is your choice. Please submit the design document written in
+markdown as a Pull Request to allow us to provide you feedback on the proposed
+design.
+
+A few notes about the design document:
+
+* Try to get the design document approved within the first 2-3 days. This is to
```
Comment (tigrato) [inline challenges/ai-ml/challenge-1.md:52]: I’m not sure this aligns well with the goals of the data analyst challenge. Candidates should first spend time exploring the dataset independently-understanding its structure, identifying patterns, and discovering what’s meaningful. Requiring the RFD to be approved within the first 2–3 days feels counterintuitive to a typical data analysis process.

File: challenges/ai-ml/challenge-1.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+### Testing
+
+Key components of the challenge should have tests that cover the happy and
+unhappy scenarios. Do not try to achieve 100% test coverage as that will take
+too long.
+
+### Dependencies
+
+Please write as much of your own code as possible. Avoid relying on third
```
Comment (tigrato) [inline challenges/ai-ml/challenge-1.md:101]: We should also allow them to use pandas and other DS libraries. Maybe consider removing it

File: challenges/ai-ml/challenge-1.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+feel is reasonable with it. In the scenario where you don't pass, you can open
+source it with any license and use it as a portfolio project.
+
+## Areas of focus
+
+Teleport focuses on networking, infrastructure and security.
+
+These are the areas we will be evaluating in the submission:
+
+* Use consistent coding style. We follow
```
Comment (tigrato) [inline challenges/ai-ml/challenge-1.md:147]: candidates will write this using python, no need to refer the go style

File: challenges/ai-ml/challenge-1.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+  and making many mistakes in the caching layer validation logic. Not having
+  caching would have solved this problem.
+* Data races. We will scan the code with a race detector and do our best to find
+  data races in the code. Avoid global state as much as possible; if using
+  global state, write down a good description why it is necessary and protect it
+  against data races.
+* Deadlocks. When using mutexes, channels or any other synchronization
+  primitives, make sure the system won't deadlock. We've seen candidates' code
+  holding a mutex and making a network call without timeouts in place. Be extra
+  careful with networking and sync primitives.
```
Comment (tigrato) [inline challenges/ai-ml/challenge-1.md:208]:
```suggestion
```
we should drop this as well

File: challenges/ai-ml/challenge-1.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+* Unstructured code. We've seen candidates leaving commented chunks of code,
+  having one large file with all the code, not having code structure at all.
+* Not communicating. Some candidates have submitted all their code to the master
+  branch without raising pull requests, which does not give us the ability to
+  provide feedback on the various implementation phases. We are a distributed
+  team, so structured, asynchronous communication is critical to us.
+* Implementing custom security algorithms/authentication schemes is always a bad
+  idea unless you are a trained security researcher/engineer. It is definitely a
+  bad idea for this task - try to stick to industry proven security methods as
+  much as possible.
```
Comment (tigrato) [inline challenges/ai-ml/challenge-1.md:218]:
```suggestion
```

File: challenges/ai-ml/challenge-1.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+## Questions
```
Comment (tigrato) [inline challenges/ai-ml/challenge-1.md:220]: This section doesn't seem to apply

File: challenges/ai-ml/challenge-1.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+permissions probing and brute-force attempts, unusual API call frequencies.
+
+After analyzing the logs, the tool should output the following information:
+
+- Whether the suspicious activity was detected and the confidence level.
+- The suspected identity.
+- The time frame of the suspected recon activity.
+- Key indicators (example logs/API calls).
+
+Include a README documentation with the tool.
```
Comment (tigrato) [inline challenges/ai-ml/challenge-1.md:130]: Should we ask them to implement the best approach for this or any approach is valid?
I am wondering because we say to keep the challenge simple but in reality we want candidates to apply the best possible approach

File: challenges/ai-ml/challenge-1.md (hunk 3 above)
Comment (strideynet) [reply challenges/ai-ml/challenge-1.md:101]: Yeah, I'd expect to see the likes of Pandas/Scikit/Pytorch/Tensorflow being leveraged pretty heavily for this task.

**Review by russjones — APPROVED** [review]
